http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/sam/SamServiceInterp.scala

```diff
@@ -135,4 +159,130 @@ class SamServiceInterp[F[_]](apiClientProvider: SamApiClientProvider[F],
+    ev: Ask[F, AppContext]
+  ): F[Boolean] =
+    for {
+      ctx <- ev.ask
+      resourcesApi <- apiClientProvider.resourcesApi(bearerToken)
+      isAuthorized <- SamRetry
+        .retry(
+          resourcesApi.resourcePermissionV2(samResourceId.resourceType.asString, samResourceId.resourceId, action),
+          "resourcePermissionV2"
+        )
+        .adaptError { case e: ApiException =>
+          SamException.create("Error checking resource permission in Sam", e, ctx.traceId)
+        }
+
+    } yield isAuthorized
+
+  override def checkAuthz(bearerToken: String, samResourceId: SamResourceId, action: String)(implicit
+    ev: Ask[F, AppContext]
+  ): F[Unit] = for {
+    ctx <- ev.ask
+    isAuthorized <- isAuthorized(bearerToken, samResourceId, action)
+    userEmail <- getUserEmail(bearerToken)
+    _ <- F.raiseWhen(!isAuthorized)(
+      SamException.create(
+        s"User $userEmail is not authorized to perform action $action on ${samResourceId.resourceType} ${samResourceId.resourceId}",
+        StatusCodes.Forbidden.intValue,
+        ctx.traceId
+      )
+    )
+    _ <- logger.info(ctx.loggingCtx)(
+      s"User $userEmail is authorized to $action ${samResourceId.resourceType} ${samResourceId.resourceId}"
+    )
+  } yield ()
+
+  override def listResources(bearerToken: String, samResourceType: SamResourceType)(implicit
+    ev: Ask[F, AppContext]
+  ): F[List[String]] = for {
+    ctx <- ev.ask
+    resourcesApi <- apiClientProvider.resourcesApi(bearerToken)
+    resources <- SamRetry
+      .retry(resourcesApi.listResourcesAndPoliciesV2(samResourceType.asString), "listResourcesAndPoliciesV2")
+      .adaptError { case e: ApiException =>
+        SamException.create("Error listing resources from Sam", e, ctx.traceId)
+      }
+  } yield resources.asScala.toList.map(_.getResourceId)
+
+  override def createResource(bearerToken: String,
+                              samResourceId: SamResourceId,
+                              projectParent: Option[GoogleProject],
+                              workspaceParent: Option[WorkspaceId],
+                              creator: Option[WorkbenchEmail]
+  )(implicit
+    ev: Ask[F, AppContext]
+  ): F[Unit] =
+    for {
+      ctx <- ev.ask
+      resourcesApi <- apiClientProvider.resourcesApi(bearerToken)
+
+      // Parent must be google project or workspace, but not both
+      parent <- (projectParent, workspaceParent) match {
+        case (Some(project), None)   => F.pure(ProjectSamResourceId(project))
+        case (None, Some(workspace)) => F.pure(WorkspaceResourceSamResourceId(workspace))
+        case _ =>
+          F.raiseError(
+            LeoInternalServerError(
+              "Internal error creating Sam resource: google project or workspace parent is required",
+              Some(ctx.traceId)
+            )
+          )
+      }
+
+      // All Leo resources have a creator role
```

Comment:
Is this true for apps with the workspace-shared access policy? Leo's config has an "owner" role there. I'm assuming also that when you say "leo resource" you're referring only to v1 stuff? I'm not sure how WSM handles these roles.

Reply:
Also, are there cases where the Leonardo SA is creating resources?

Reply:
So I was trying to assume "desired state" here, where we migrate away from WSM-owned resources, and use dedicated resource types for Leo. That was a recommendation in [this doc](https://docs.google.com/document/d/1FAkxK25hqCdCJWXVdcmpaTE23IeTPHSm_lyR_MuoMf8/edit).

Hm you're right that `kubernetes-app-shared` has `owner`/`user` roles. And `kubernetes-app` has `creator`/`manager`.

Maybe I'll revisit the method signature here… I was trying to see if `createResource` could be generic to all types of resources Leo supports, or if we need a `createRuntimeResource`, `createDiskResource`, `createKubernetesAppResource`, etc.

Reply:
That makes sense. And I love the idea of a standardized Leo-managed resource with predictable roles. Maybe creating a resource gives you the Creator role but depending on the resource, the context, and the user, they could get other roles as well? This might just require a bit more design to nail down an approach.

Reply:
We could start with this method and add other methods for special cases if we need to.

Reply:
@mlilynolting what do you think of [this change](https://github.com/DataBiosphere/leonardo/pull/4779/commits/a74f5bfc48fca912b37913856d0cbc2ec467bb1e)? Instead of assuming a `creator` role, I updated `SamService.createResource` to just take the policies directly. We can maybe add specialized methods (`getRuntimePolicies`, `getAppPolicies`, etc) later on.

Reply:
I think that's a solid change that reflects the Sam contract really faithfully - Leo calls Sam with a new policy it wants to create for this resource. The policy name doesn't need to be unique, right?